classes/affiliation/Repository.php

```diff
@@ -0,0 +1,302 @@
+            foreach ($affiliations as $affiliation) {
+            $affiliation['authorId'] = $props['id'];
+
+                $validator = ValidatorFactory::make(
+                    $affiliation,
+                    $schemaService->getValidationRules($this->dao->schema, $allowedLocales)
+                );
+
+//                // Check for input from disallowed locales
+//                ValidatorFactory::allowedLocales(
+//                    $validator,
+//                    $this->schemaService->getMultilingualProps(PKPSchemaService::SCHEMA_AFFILIATION),
+//                    $allowedLocales
+//                );
+
+                // The ror or one name must exist
+                $validator->after(function ($validator) use ($affiliation) {
+                    if (empty($affiliation['ror']) && empty($affiliation['name'])) {
+                        $validator->errors()->add('affiliations-affiliationId', __('author.affiliationRorAndNameEmpty'));
+                    }
+                });
+
+                if ($validator->fails()) {
+                    $errors = $this->schemaService->formatValidationErrors($validator->errors());
+                    break;
+                }
+            }
+        }
+
+        Hook::call('Affiliation::validate', [&$errors, $author, $props, $submission, $context]);
+
+        return $errors;
+    }
+
+    /** @copydoc DAO::insert() */
+    public function add(Affiliation $affiliation): int
+    {
+        $id = $this->dao->insert($affiliation);
+        Hook::call('Affiliation::add', [$affiliation]);
+        return $id;
+    }
+
+    /** @copydoc DAO::update() */
+    public function edit(Affiliation $affiliation, array $params): void
+    {
+        $newRow = clone $affiliation;
+        $newRow->setAllData(array_merge($newRow->_data, $params));
+        Hook::call('Affiliation::edit', [$newRow, $affiliation, $params]);
+        $this->dao->update($newRow);
+    }
+
+    /** @copydoc DAO::delete() */
+    public function delete(Affiliation $affiliation): void
+    {
+        Hook::call('Affiliation::delete::before', [$affiliation]);
+        $this->dao->delete($affiliation);
+        Hook::call('Affiliation::delete', [$affiliation]);
+    }
+
+    /**
+     * Delete a collection of affiliations
+     */
+    public function deleteMany(Collector $collector): void
+    {
+        foreach ($collector->getMany() as $row) {
+            $this->delete($row);
+        }
+    }
+
+    /**
+     * Get all affiliations for a given author.
+     */
+    public function getByAuthorId(int $authorId, ?string $submissionLocale = null): array
+    {
+        return iterator_to_array(
+            $this->getCollector()
+                ->filterByAuthorId($authorId)
+                ->getMany($submissionLocale)
+        );
+    }
+
+    /**
+     * Save affiliations.
+     */
+    public function saveAffiliations(Author $author): void
+    {
+        $affiliations = $author->getAffiliations();
```

Comment:
Here we can expect the affiliations to always be an array of Affiliation objects. (I commented above in PKPSubmissionController in add/editContributor)

Reply:
There are many places where Repo::author()->add() and Repo::author()->edit() are called. If I don't do the check here, I expect that there will be unforeseen errors. That's why I added the extra check here.